playground/frontend/playground_components/lib/src/widgets/run_button.dart

```diff
@@ -31,67 +32,100 @@ const _width = 150.0;
 
 class RunButton extends StatelessWidget {
   final PlaygroundController playgroundController;
-  final bool isRunning;
   final VoidCallback runCode;
   final VoidCallback cancelRun;
   final bool disabled;
 
   const RunButton({
     super.key,
     required this.playgroundController,
-    required this.isRunning,
     required this.runCode,
     required this.cancelRun,
     this.disabled = false,
   });
 
   @override
   Widget build(BuildContext context) {
-    return SizedBox(
-      width: _width,
-      height: BeamSizes.buttonHeight,
-      child: ShortcutTooltip(
-        shortcut: playgroundController.runShortcut,
-        child: ElevatedButton.icon(
-          style: const ButtonStyle(
-            padding: MaterialStatePropertyAll(EdgeInsets.zero),
+    return AnimatedBuilder(
+      animation: playgroundController.codeRunner,
+      builder: (context, child) {
+        final isRunning = playgroundController.codeRunner.isCodeRunning;
+        final runStartDate = playgroundController.codeRunner.runStartDate;
+
+        return SizedBox(
+          width: _width,
+          height: BeamSizes.buttonHeight,
+          child: ShortcutTooltip(
+            shortcut: playgroundController.runShortcut,
+            child: ElevatedButton.icon(
+              style: const ButtonStyle(
+                padding: MaterialStatePropertyAll(EdgeInsets.zero),
+              ),
+              icon: isRunning
+                  ? SizedBox(
+                      width: BeamIconSizes.small,
+                      height: BeamIconSizes.small,
+                      child: CircularProgressIndicator(
+                        color: Theme.of(context)
+                            .extension<BeamThemeExtension>()!
+                            .primaryBackgroundTextColor,
+                      ),
+                    )
+                  : const Icon(Icons.play_arrow),
+              label: isRunning
+                  ? PeriodicRebuilderWidget(
+                      rebuildFrequencyMls: 1,
+                      builder: () {
+                        return _ButtonText(
+                          isRunning: isRunning,
+                          runStartDate: runStartDate,
+                        );
+                      },
+                    )
+                  : _ButtonText(
+                      isRunning: isRunning,
+                      runStartDate: runStartDate,
+                    ),
+              onPressed: _onPressHandler(),
```

Comment:
It's easy to get confused when returning callbacks, so
```suggestion
              onPressed: disabled ? null : _onPressed,
```
and handle it there.